firebase-firestore/src/androidTest/java/com/google/firebase/firestore/ValidationTest.java

```diff
@@ -432,6 +435,56 @@ public void queriesCannotBeCreatedFromDocumentsMissingSortValues() {
     expectError(() -> query.endAt(snapshot), reason);
   }
 
+  @Test
+  public void queriesCannotBeSortedByAnUncommittedServerTimestamp() {
+    CollectionReference collection = testCollection();
+
+    // Ensure the server timestamp stays uncommitted for the first half of the test
+    waitFor(collection.firestore.getClient().disableNetwork());
+
+    TaskCompletionSource<Void> offlineCallbackDone = new TaskCompletionSource<>();
+    TaskCompletionSource<Void> onlineCallbackDone = new TaskCompletionSource<>();
+
+    collection.addSnapshotListener(
+        (snapshot, error) -> {
+          assertNotNull(snapshot);
+
+          // Skip the initial empty snapshot.
+          if (snapshot.isEmpty()) return;
+
+          assertThat(snapshot.getDocuments()).hasSize(1);
+          DocumentSnapshot docSnap = snapshot.getDocuments().get(0);
+
+          if (snapshot.getMetadata().hasPendingWrites()) {
+            // Offline snapshot. Since the server timestamp is uncommitted, we shouldn't be able to
+            // query by it.
+            assertThrows(
+                IllegalArgumentException.class,
+                () ->
+                    collection
+                        .orderBy("timestamp")
+                        .endAt(docSnap)
+                        .addSnapshotListener((snapshot2, error2) -> {}));
```

Comment:
Is `addSnapshotListener` necessary?

Reply:
Not strictly. It *is* necessary to trigger the internal error that would otherwise occur though. For that reason, I'd like to keep the snapshot listener in the 'online' case just a few lines down. And if I'm keeping it there, then keeping it here for symmetry seems to make some sense. (Though I don't mind removing it either.) wdyt?

Reply:
Sounds good.